### ISSUE TITLE
Add support for HTTP2 prio and dependencies (take 2)

### ISSUE THREAD

**Issue by bagder** [bagder/curl#499]
The current code lives in this branch: https://github.com/bagder/curl/tree/http2-priodeps (rebased occasionally)

Basic testing seems to imply it works!

@tatsuhiro-t: I'll appreciate if you'd give it a look and see if you can catch any major problems this time around.

**Comment by tatsuhiro-t** [comment]
I was going to try this out, but instead got this assertion failure:
```
a.out: http2.c:429: int on_frame_recv(nghttp2_session *, const nghttp2_frame *, void *): Assertion `conn->data == data_s' failed.
```
I use 2 concurrent streams.
```
Connection state changed (HTTP/2 confirmed)
TCP_NODELAY set
Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
nghttp2_session_mem_recv() returns 0
http2_send len=53
Using Stream ID: 1 (easy handle 0x62d00000a408)
before_frame_send() was called
on_frame_send() was called, length = 20
STATE: DO => DO_DONE handle 0x62d00000a408; line 1350 (connection #0) 
multi changed, check CONNECT_PEND queue!
STATE: CONNECT_PEND => CONNECT handle 0x62d000000408; line 2791 (connection #-5000) 
STATE: DO_DONE => WAITPERFORM handle 0x62d00000a408; line 1477 (connection #0) 
STATE: WAITPERFORM => PERFORM handle 0x62d00000a408; line 1487 (connection #0) 
http2_recv: 16384 bytes buffer at 0x629000000208 (stream 1)
nread=15
nghttp2_session_mem_recv() returns 15
All data in connection buffer processed
http2_recv returns AGAIN for stream 1
Found bundle for host localhost: 0x60400001bb58
Conn: 0 (0x61c00000f888) Receive pipe weight: (-1/0), penalized: FALSE
Multiplexed connection found!
Found connection 0, with requests in the pipe (1)
Re-using existing connection! (#0) with host localhost
STATE: CONNECT => WAITDO handle 0x62d000000408; line 1149 (connection #0) 
STATE: WAITDO => DO handle 0x62d000000408; line 1303 (connection #0) 
http2_send len=53
Using Stream ID: 3 (easy handle 0x62d000000408)
before_frame_send() was called
on_frame_send() was called, length = 5
STATE: DO => DO_DONE handle 0x62d000000408; line 1350 (connection #0) 
STATE: DO_DONE => WAITPERFORM handle 0x62d000000408; line 1477 (connection #0) 
STATE: WAITPERFORM => PERFORM handle 0x62d000000408; line 1487 (connection #0) 
http2_recv: 16384 bytes buffer at 0x629000000208 (stream 1)
nread=161
on_begin_headers() was called
h2 status: HTTP/2 200
h2 header: server: nghttpd nghttp2/1.3.5-DEV
h2 header: content-length: 10
h2 header: cache-control: max-age=3600
h2 header: date: Tue, 20 Oct 2015 14:43:15 GMT
h2 header: last-modified: Tue, 30 Sep 2014 12:40:52 GMT
on_frame_recv() header 1 stream 1
Store 178 bytes headers from stream 1 at 0x629000000208
on_begin_headers() was called
h2 status: HTTP/2 200
h2 header: server: nghttpd nghttp2/1.3.5-DEV
h2 header: content-length: 10
h2 header: cache-control: max-age=3600
h2 header: date: Tue, 20 Oct 2015 14:43:15 GMT
h2 header: last-modified: Tue, 30 Sep 2014 12:40:52 GMT
on_frame_recv() header 1 stream 3
a.out: http2.c:429: int on_frame_recv(nghttp2_session *, const nghttp2_frame *, void *): Assertion `conn->data == data_s' failed.
```

Not sure why this fails. Same code worked before.

**Comment by tatsuhiro-t** [comment]
I have to say that this also occur in master branch..

**Comment by tatsuhiro-t** [comment]
If I commented out http2.c line 429, then it works.

**Comment by tatsuhiro-t** [comment]
It seems working fine with new options.  I only tested with HEADERS frame though.

I noticed we call weight as priority or prio in curl API.  I prefer weight to prio here, because this is not priority, and the ratio of the resource allocated to the parent stream.

**Comment by bagder** [comment]
Thanks for your valuable feedback, I'll need to investigate the assert thing.

I agree with your point about prio vs weight. I originally thought that I could design the API to be more generic to not necessarily always have to be about HTTP/2 but I think I'm making us a disservice by that and I'll switch the terminology over to "weight" before I merge this to master.

**Comment by bagder** [comment]
I've removed the asserts (I deem them incorrect) and I renamed the option to use "weight" now - a rebased version has been pushed. I'll let it sit in its own branch for a few more days just in case someone else gets around to play with it and can give it a test before I merge it to master.

**Comment by bagder** [comment]
Merged in master as of commit d31ed6ac719

**Comment by patrikhson** [comment]
Hi,

Yesterday I  read some news and first I was surprised, then I  realized it can be true, please read it and tell me your thoughts  <http://thokethumu.acousticar.com/xybhisw>

Hugs, paf@frobbit.se
